### PR TITLE
Add support for varmap

### LINKED ISSUE #45: vars should be able to take non-flat values
### What happened?
Usecase: We are trying to create EKS Cluster using Crossplane and to create multiple Node-groups/launch-templates we are using terraform-provider to dynamically read the details from the claim.yaml

claim.yaml

```apiVersion: abc.io/v1alpha1
kind: ManagedCluster
metadata:
  namespace: crossplane-system
  name: eks-cluster
parameters:
    region: us-east-2
    version: "1.21"
    nodeRole: arn:aws:iam::xxxxx:role/eksNodeRole
    nodeGroups:
      - desiredSize: 2
        maxSize: 2
        minSize: 2
        instanceType: m5.4xlarge
        label: fruit=apple
        volumeSize: 150
      - desiredSize: 1
        maxSize: 1
        minSize: 1
        instanceType: m5.4xlarge
        label: fruit=banana
        volumeSize: 80
```

As per current behavior, terraform-provider gives 2 options to pass variables - `vars` & `varFiles`. 

`varFiles` supports data in HCL or JSON format.
`vars` support data in plain string - flat values only.

### What environment did it happen in?
Crossplane version: [1.10.1](https://github.com/crossplane/crossplane/releases/tag/v1.10.1)
provider-terraform: [0.2.0](https://marketplace.upbound.io/providers/upbound/provider-terraform/v0.2.0)
Kubernetes: 1.23

Solution could be to enhance `vars` to take non-flat values as input & internally convert to HCL/JSON to pass to terraform-provider.

### PR DESCRIPTION
**Pull request by nakamume**
## Description of your changes
Fixes #45

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

## How has this code been tested

- I have run the unit tests.
### E2E test with following configs
`provider-config`
```
apiVersion: tf.upbound.io/v1beta1
kind: ProviderConfig
metadata:
  name: default
spec:
  credentials:
  - source: None
    filename: xyz.json
  configuration: |
    terraform {
      backend "kubernetes" {
        secret_suffix     = "providerconfig-default"
        namespace         = "default"
        in_cluster_config = false
      }
    }
```

`workspace` object
```
apiVersion: tf.upbound.io/v1beta1
kind: Workspace
metadata:
  name: random-string
  annotations:
    crossplane.io/external-name: test
spec:
  providerConfigRef:
    name: default
  forProvider:
    source: Inline
    vars:
    - key: length
      value: "16"
    varmap:
      config:
        lower: true
        upper: true
        numeric: true
        special: false
    module: |
      variable length {}
      variable config {}

      resource "random_string" "random" {
        length  = tonumber(var.length)
        lower   = var.config.lower
        upper   = var.config.upper
        numeric = var.config.numeric
        special = var.config.special
      }

      output random {
        value = random_string.random
      }
```

Resulted in following
A tfvar file `crossplane-provider-terraform-0.tfvars.json` with content
```
{"config":{"lower":true,"numeric":true,"special":false,"upper":true}}
```

Description of the workspace object
```
kubectl describe workspace.tf.upbound.io/random-string

Name:         random-string
Namespace:    
Labels:       <none>
Annotations:  crossplane.io/external-create-pending: 2023-03-10T12:35:19+09:00
              crossplane.io/external-create-succeeded: 2023-03-10T12:35:24+09:00
              crossplane.io/external-name: test
API Version:  tf.upbound.io/v1beta1
Kind:         Workspace
Metadata:
  Creation Timestamp:  2023-03-10T03:35:11Z
  Finalizers:
    finalizer.managedresource.crossplane.io
  Generation:  1
  Resource Version:  2602
  UID:               654eec20-7cd4-473f-8de7-1f73bb85438d
Spec:
  Deletion Policy:  Delete
  For Provider:
    Entrypoint:  
    Module:      variable length {}
variable config {}

resource "random_string" "random" {
  length  = tonumber(var.length)
  lower   = var.config.lower
  upper   = var.config.upper
  numeric = var.config.numeric
  special = var.config.special
}

output random {
  value = random_string.random
}

    Source:  Inline
    Varmap:
      Config:
        Lower:    true
        Numeric:  true
        Special:  false
        Upper:    true
    Vars:
      Key:    length
      Value:  16
  Provider Config Ref:
    Name:  default
Status:
  At Provider:
    Outputs:
      Random:  {"id":"ykCoPq5NRFaNXRjW","keepers":null,"length":16,"lower":true,"min_lower":0,"min_numeric":0,"min_special":0,"min_upper":0,"number":true,"numeric":true,"override_special":null,"result":"ykCoPq5NRFaNXRjW","special":false,"upper":true}
  Conditions:
    Last Transition Time:  2023-03-10T03:35:35Z
    Reason:                Available
    Status:                True
    Type:                  Ready
    Last Transition Time:  2023-03-10T03:35:24Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced
Events:
  Type    Reason                   Age   From                             Message
  ----    ------                   ----  ----                             -------
  Normal  CreatedExternalResource  6m3s  managed/workspace.tf.upbound.io  Successfully requested creation of external resource

```

### CODE REVIEW

**Comment by bobh66** [comment]
Hi @nakamume - thanks for the PR.  Can you also update README.md with an example?  Thanks

**Review by ytsarev — CHANGES_REQUESTED** [review]
Please provide a full description according to PR template with a special focus on the testing part.

File: internal/controller/workspace/workspace.go [hunk 1]
```diff
@@ -420,6 +422,14 @@ func (c *external) options(ctx context.Context, p v1beta1.WorkspaceParameters) (
 		}
 	}
 
+	if p.VarMap != nil {
```
[the remark this answers is not in the document]
Comment (nakamume) [reply internal/controller/workspace/workspace.go:426]: Done here: https://github.com/upbound/provider-terraform/pull/112/commits/f40accab4c3d98222acfbdaecf828885b35d6111

**Comment by nakamume** [comment]
@bobh66 @ytsarev - addressed your comments. PTAL

**Review by ytsarev — APPROVED** [review]
Looks great, thanks for the updates!

**Review by bobh66 — APPROVED** [review]
LGTM - thanks!

**Pull request merged**